File: experiment-15-sonnet5/brazil/runs/language=python_model=sonnet-5_prompt=none/rep1/soccer_mcp/team_names.py
"""Team name normalization.

The source CSVs spell the same club in several incompatible ways: with a
state suffix ("Palmeiras-SP"), with a country suffix ("Barcelona-EQU"),
with a parenthetical qualifier ("Guaraní (PAR)"), with old/new brand names
("Atletico Paranaense" vs "Athletico-PR"), or as a bare short name
("Flamengo"). Matching across datasets requires collapsing all of these to
one canonical key while still keeping a readable display name.
"""

from __future__ import annotations

import re
import unicodedata

# Trailing " - XX" / "-XX" state or country qualifiers, e.g. "Flamengo-RJ",
# "Boca Juniors - ARG", "Barcelona-EQU". Requires a dash so we don't clip
# legitimate short words.
_SUFFIX_RE = re.compile(r"\s*-\s*[A-Za-z]{2,4}$")
# A trailing parenthetical short code, e.g. "Guaraní (PAR)". This is a
# qualifier just like a dash-suffix (it disambiguates same-named clubs from
# different countries) so it must be preserved, not discarded like a
# descriptive parenthetical aside.
_PAREN_CODE_RE = re.compile(r"\(\s*([A-Za-z]{2,4})\s*\)\s*$")
_PAREN_RE = re.compile(r"\s*\([^)]*\)")
_WS_RE = re.compile(r"\s+")


def _paren_code_to_suffix(text: str) -> str:
    match = _PAREN_CODE_RE.search(text)
    if not match:
        return text
    return f"{text[:match.start()].rstrip()}-{match.group(1)}"

# Maps a generic-normalized variant -> the generic-normalized canonical key.
# Only needed where different datasets use genuinely different words for the
# same club (rebrands, abbreviations, alternate short forms). State-suffix
# and punctuation differences are already handled by the generic normalizer.
_ALIASES: dict[str, str] = {
    # Athletico Paranaense rebranded from "Atletico Paranaense" in 2018.
    "atletico paranaense": "athletico paranaense",
    "athletico paranaense": "athletico paranaense",
    "atletico pr": "athletico paranaense",
    "athletico pr": "athletico paranaense",
    "athletico": "athletico paranaense",
    # Atletico Mineiro / Atletico-MG
    "atletico mineiro": "atletico mg",
    # Sport Club do Recife
    "sport recife": "sport",
    "sport club do recife": "sport",
    "sport pe": "sport",
    # Nautico
    "nautico pe": "nautico",
    "clube nautico capibaribe": "nautico",
    # America-MG
    "america mg": "america mg",
    "america fc": "america mg",
    "america fc minas gerais": "america mg",
    "america futebol clube mg": "america mg",
    # Vasco da Gama
    "vasco": "vasco da gama",
    "vasco rj": "vasco da gama",
    "clube de regatas vasco da gama": "vasco da gama",
    # Sao Paulo
    "sao paulo sp": "sao paulo",
    "sao paulo futebol clube": "sao paulo",
    # Corinthians
    "corinthians sp": "corinthians",
    "sport club corinthians paulista": "corinthians",
    # Botafogo (Rio) - do not merge with Botafogo-PB / Botafogo-SP.
    "botafogo rj": "botafogo",
    # Red Bull Bragantino
    "red bull bragantino sp": "red bull bragantino",
    "bragantino sp": "red bull bragantino",
    "rb bragantino": "red bull bragantino",
    # Cross-state abbreviated forms seen in Brasileirao_Matches.csv
    "csa al": "csa",
    "cuiaba mt": "cuiaba",
    "goias go": "goias",
    "bahia ba": "bahia",
    "vitoria ba": "vitoria",
    "ceara ce": "ceara",
    "fortaleza ce": "fortaleza",
    "avai sc": "avai",
    "chapecoense sc": "chapecoense",
    "coritiba pr": "coritiba",
    "criciuma sc": "criciuma",
    "cruzeiro mg": "cruzeiro",
    "figueirense sc": "figueirense",
    "gremio rs": "gremio",
    "internacional rs": "internacional",
    "joinville sc": "joinville",
    "juventude rs": "juventude",
    "parana pr": "parana",
    "ponte preta sp": "ponte preta",
    "portuguesa sp": "portuguesa",
    "santa cruz pe": "santa cruz",
    "santos sp": "santos",
    "flamengo rj": "flamengo",
    "fluminense rj": "fluminense",
    "palmeiras sp": "palmeiras",
    "atletico go": "atletico go",
}

# Nice display names for canonical keys (accents restored, standard casing).
_CANONICAL_DISPLAY: dict[str, str] = {
    "athletico paranaense": "Athletico Paranaense",
    "atletico mg": "Atletico-MG",
    "atletico go": "Atletico-GO",
    "sport": "Sport",
    "nautico": "Nautico",
    "america mg": "America-MG",
    "vasco da gama": "Vasco da Gama",
    "sao paulo": "Sao Paulo",
    "corinthians": "Corinthians",
    "botafogo": "Botafogo",
    "red bull bragantino": "Red Bull Bragantino",
    "csa": "CSA",
    "cuiaba": "Cuiaba",
    "goias": "Goias",
    "bahia": "Bahia",
    "vitoria": "Vitoria",
    "ceara": "Ceara",
    "fortaleza": "Fortaleza",
    "avai": "Avai",
    "chapecoense": "Chapecoense",
    "coritiba": "Coritiba",
    "criciuma": "Criciuma",
    "cruzeiro": "Cruzeiro",
    "figueirense": "Figueirense",
    "gremio": "Gremio",
    "internacional": "Internacional",
    "joinville": "Joinville",
    "juventude": "Juventude",
    "parana": "Parana",
    "ponte preta": "Ponte Preta",
    "portuguesa": "Portuguesa",
    "santa cruz": "Santa Cruz",
    "santos": "Santos",
    "flamengo": "Flamengo",
    "fluminense": "Fluminense",
    "palmeiras": "Palmeiras",
}


def strip_accents(text: str) -> str:
    normalized = unicodedata.normalize("NFKD", text)
    return "".join(c for c in normalized if not unicodedata.combining(c))


def _strip_suffix_and_noise(raw: str) -> str:
    text = raw.strip()
    text = _PAREN_RE.sub("", text)
    text = _SUFFIX_RE.sub("", text)
    text = text.replace(".", "")
    return _WS_RE.sub(" ", text).strip(" -")


def normalize_team(raw: str) -> tuple[str, str]:
    """Return (canonical_key, display_name) for a raw team name.

    canonical_key is a lowercase, accent-free string that is stable across
    all datasets for the same club. display_name is a human-readable label.

    Two candidate keys are tried against the alias table: one that keeps the
    trailing state/country code as a word (so e.g. "Atletico - MG" and
    "Atletico - PR" stay distinguishable) and one with the suffix removed
    entirely. The suffix-preserving key is tried first since it disambiguates
    same-named clubs from different states.
    """
    if raw is None:
        return "", ""
    with_suffix = _paren_code_to_suffix(str(raw).strip())
    no_paren = _PAREN_RE.sub("", with_suffix).strip()
    ascii_lower = strip_accents(no_paren).lower().replace(".", "")
    suffix_key = _WS_RE.sub(" ", ascii_lower.replace("-", " ")).strip()
    stripped_key = _WS_RE.sub(" ", _SUFFIX_RE.sub("", ascii_lower)).strip(" -")

    # Prefer the alias table, then the suffix-preserving key (safer default:
    # it never merges two different states' clubs of the same short name),
    # falling back to the fully-stripped key only if that's all we have.
    canonical_key = _ALIASES.get(suffix_key) or _ALIASES.get(stripped_key) or suffix_key or stripped_key

    display = _CANONICAL_DISPLAY.get(canonical_key)
    if display is None:
        cleaned = _strip_suffix_and_noise(no_paren)
        display = cleaned if cleaned else str(raw).strip()
    return canonical_key, display


def teams_match(raw_a: str, raw_b: str) -> bool:
    return normalize_team(raw_a)[0] == normalize_team(raw_b)[0]

File: experiments/adrianco/experiment-5/runs/language=java_model=claude-opus-4-7_tooling=none/rep3/src/main/java/com/soccer/mcp/util/TeamNames.java
package com.soccer.mcp.util;

import java.text.Normalizer;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Team-name normalisation.
 *
 * Source CSVs use different naming conventions:
 *  - With state suffix: "Palmeiras-SP", "Flamengo-RJ"
 *  - Without suffix:    "Palmeiras", "Flamengo"
 *  - Full names:        "Sport Club Corinthians Paulista"
 *  - Country suffix:    "Nacional (URU)"
 *
 * For some teams (Flamengo, Palmeiras, ...) the short form is unique and the
 * suffix can safely be stripped. For others ("Atletico-MG" vs "Atletico-PR")
 * the suffix is the only disambiguator. We therefore use an explicit alias map
 * keyed on a stripped+lowercased form, and only fall back to the raw stripped
 * form when no alias is found.
 */
public final class TeamNames {

    private static final Pattern COUNTRY_PAREN = Pattern.compile("\\s*\\(([A-Za-z]{2,4})\\)\\s*$");
    private static final Pattern WHITESPACE = Pattern.compile("\\s+");

    /**
     * Alias map. Keys are pre-normalized forms (accent-stripped, lowercased,
     * whitespace-collapsed). Values are the canonical name we want to dedupe to.
     * Both "with-suffix" and "without-suffix" forms map to the same canonical.
     */
    private static final Map<String, String> ALIASES = buildAliases();

    private TeamNames() {}

    private static Map<String, String> buildAliases() {
        Map<String, String> m = new HashMap<>();
        // Teams that appear both with and without state suffix — alias both forms.
        alias(m, "flamengo",       "flamengo", "flamengo-rj");
        alias(m, "fluminense",     "fluminense", "fluminense-rj");
        alias(m, "palmeiras",      "palmeiras", "palmeiras-sp",
                "se palmeiras", "sociedade esportiva palmeiras");
        alias(m, "corinthians",    "corinthians", "corinthians-sp",
                "sport club corinthians paulista");
        alias(m, "santos",         "santos", "santos-sp", "santos fc");
        alias(m, "sao paulo",      "sao paulo", "sao paulo-sp", "sao paulo fc");
        alias(m, "gremio",         "gremio", "gremio-rs");
        alias(m, "internacional",  "internacional", "internacional-rs", "sc internacional", "inter");
        alias(m, "vasco",          "vasco", "vasco-rj", "vasco da gama", "vasco da gama-rj");
        alias(m, "botafogo",       "botafogo", "botafogo-rj");
        alias(m, "cruzeiro",       "cruzeiro", "cruzeiro-mg");
        alias(m, "bahia",          "bahia", "bahia-ba", "ec bahia");
        alias(m, "fortaleza",      "fortaleza", "fortaleza-ce");
        alias(m, "ceara",          "ceara", "ceara-ce");
        alias(m, "coritiba",       "coritiba", "coritiba-pr");
        alias(m, "chapecoense",    "chapecoense", "chapecoense-sc");
        alias(m, "goias",          "goias", "goias-go");
        alias(m, "avai",           "avai", "avai-sc");
        alias(m, "csa",            "csa", "csa-al");
        alias(m, "atletico mineiro", "atletico-mg", "atletico mineiro", "clube atletico mineiro");
        alias(m, "atletico paranaense", "atletico-pr", "athletico-pr",
                "athletico paranaense", "atletico paranaense");
        alias(m, "atletico goianiense", "atletico-go", "athletico-go",
                "atletico goianiense", "athletico goianiense");
        alias(m, "america mineiro", "america-mg", "america mineiro");
        // Note: do NOT alias raw "atletico" without state — too ambiguous.
        return m;
    }

    private static void alias(Map<String, String> m, String canonical, String... keys) {
        for (String k : keys) m.put(k, canonical);
    }

    /** Returns a canonical lookup key suitable for dedupe and equality matching. */
    public static String normalize(String name) {
        if (name == null) return "";
        String s = name.trim();
        // Strip "(XXX)" country/competition suffix like "(URU)".
        Matcher cm = COUNTRY_PAREN.matcher(s);
        if (cm.find()) s = s.substring(0, cm.start()).trim();
        s = stripAccents(s);
        s = s.toLowerCase(Locale.ROOT);
        s = WHITESPACE.matcher(s).replaceAll(" ").trim();

        String alias = ALIASES.get(s);
        if (alias != null) return alias;

        // Fallback: drop a trailing 2-letter state code (e.g. "-rj") only if
        // there is no ambiguous mapping. We're conservative — only strip when
        // the suffix matches Brazilian state codes.
        return s;
    }

    public static String stripAccents(String s) {
        if (s == null) return null;
        String nfd = Normalizer.normalize(s, Normalizer.Form.NFD);
        return nfd.replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
    }

    /** True if either side's normalized form contains the other's. */
    public static boolean matches(String candidate, String query) {
        if (candidate == null || query == null) return false;
        String a = normalize(candidate);
        String b = normalize(query);
        if (a.equals(b)) return true;
        if (a.contains(b) || b.contains(a)) return true;
        return false;
    }
}
